benchmark/single-source/KeyPathPerformanceTests.swift

```diff
@@ -0,0 +1,1009 @@
+//===--- KeyPathPerformanceTests.swift -------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+import TestsUtils
+
+public let benchmarks = [
+  BenchmarkInfo(
+    name: "KeyPathNestedStructs",
+    runFunction: run_KeyPathNestedStructs,
+    tags: [.validation, .api],
+    setUpFunction: setupKeyPathNestedStructs
+  ),
+  BenchmarkInfo(
+    name: "KeyPathDirectAccess",
+    runFunction: run_testDirectAccess,
+    tags: [.validation, .api],
+    setUpFunction: setupSingleton
+  ),
+  BenchmarkInfo(
+    name: "KeyPathReadPerformance",
+    runFunction: run_testKeyPathReadPerformance,
+    tags: [.validation, .api],
+    setUpFunction: setupSingleton
+  ),
+  BenchmarkInfo(
+    name: "KeyPathWritePerformance",
+    runFunction: run_testKeyPathWritePerformance,
+    tags: [.validation, .api],
+    setUpFunction: setupSingleton
+  ),
+]
+
+// Number of iterations for each benchmark.
+let numberOfElementsForNestedStructs = 20000
+
+// Make it easy to switch between primitives of different types (Int, Float, Double, etc.).
+typealias ElementType = Double
+
+// Variables used across the benchmarks
+let expectedIntForNestedStructs = 3
+let arrayHolder = FixedSizeArrayHolder()
+
+let appendedKeyPath = singleHopKeyPath0.appending(path: singleHopKeyPath1)
+  .appending(path: singleHopKeyPath2).appending(path: singleHopKeyPath3)
+  .appending(path: singleHopKeyPath4)
+
+// The purpose of this class is to hold arrays used during the course of the benchmarks.
+// This is to avoid storing them in the global scope, which can slow things down.
+// We force instantiation in the setUpFunction so that we exclude the setup time from the benchmark time itself,
+// otherwise the instance would be lazily instantiated, and thereby add to the benchmark time.
+class FixedSizeArrayHolder {
+  var fixedSizeArray100: FixedSizeArray100<ElementType>
+  var mainArrayForNestedStructs: [A]
+  static let shared = FixedSizeArrayHolder()
+  init() {
+    fixedSizeArray100 = initializeFixedSizeArray100()
+    mainArrayForNestedStructs = [A]()
+  }
+
+  func reset() {
+    fixedSizeArray100 = initializeFixedSizeArray100()
+  }
+
+  func forceInstantiation() {}
```

Comment:
This does not work. In optimized code a call to this function is completely optimized away.

Reply:
Yes, I need to keep in mind that the benchmarks are run with all three of -O, -Onone, and -Osize.